### PR TITLE
Setbacks refactor

### PR DESCRIPTION
**Pull request by ppinchuk**
Major code refactor to allow either solar or wind regulations inputs to all setback types, including future setback implementations.

### CODE REVIEW

File: reVX/setbacks/structure_setbacks.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+import os
+import re
+import logging
+
+from reVX.setbacks.base import BaseSetbacks
+
+logger = logging.getLogger(__name__)
+
+
+class StructureSetbacks(BaseSetbacks):
```
Comment (grantbuster) [inline reVX/setbacks/structure_setbacks.py:14]: nice, these look like really clean + simple concrete subclasses. A sign of a good refactor!

**Review by grantbuster — CHANGES_REQUESTED** [review]

File: reVX/setbacks/base.py [hunk 2]
```diff
@@ -60,54 +60,173 @@ def features_clipped_to_county(features, cnty):
     return tmp[~tmp.is_empty]
 
 
+class _Rasterizer:
```
[the remark this answers is not in the document]
Comment (ppinchuk) [reply reVX/setbacks/base.py:63]: Yea, that's a good point actually. This is why it's good to have a second set of eyes, thank you! :)

File: reVX/setbacks/base.py [hunk 3]
```diff
@@ -60,54 +60,173 @@ def features_clipped_to_county(features, cnty):
...
+            For example, a cell that was previously excluded with a
+            a boolean mask (value of 1) may instead be converted to an
+            inclusion weight value of 0.75, meaning that 75% of the area
+            corresponding to that point should be included (i.e. the
+            exclusion feature only intersected a small portion - 25% -
+            of the cell). This percentage inclusion value is calculated
+            by upscaling the output array using this input value,
+            rasterizing the exclusion features onto it, and counting the
+            number of resulting sub-cells excluded by the feature. For
+            example, setting the value to `3` would split each output
```
[the remark this answers is not in the document]
Comment (ppinchuk) [reply reVX/setbacks/base.py:84]: Yea... it's tricky b/c it has to work this way, unless we put in a flag and then have a required value if the flag is true, which is slightly messier IMO (though let me know if you disagree and I can change it). I did try to compensate for the non-straightforward nature of this value with a through docstring, which I updated based on this comment. Let me know if you think it's still too confusing.

File: reVX/setbacks/base.py [hunk 4]
```diff
@@ -60,54 +60,173 @@ def features_clipped_to_county(features, cnty):
...
+        """Aggregate the high resolution exclusions array to output shape. """
+
+        arr = self._no_exclusions_array().astype(np.float32)
+        for i, j in product(range(self._scale_factor),
+                            range(self._scale_factor)):
+            arr += hr_arr[i::self._scale_factor, j::self._scale_factor]
+        return arr
+
+
 class BaseSetbacks:
```
[the remark this answers is not in the document]
Comment (ppinchuk) [reply reVX/setbacks/base.py:212]: Yea good call... This class actually used to be abstract but went through so many iterations that it got lost in translation. In the current state there is at least 1 function that I would call abstract so I do think it makes sense to make it an ABC. Added this change, thank you!

File: reVX/setbacks/base.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ppinchuk) [reply reVX/setbacks/base.py]: Yes, that is a reV limitation. There is no option for "exclusion weights", so if anyone wanted to model partial setbacks, they have to input inclusion weights to reV: https://github.com/NREL/reV/blob/13525b8802e526dca09cd8ee66784a9f3e01eb60/reV/supply_curve/exclusions.py#L23-L35

[the remark this answers is not in the document]
Comment (grantbuster) [reply reVX/setbacks/base.py]: Ah i see, okay thats annoying on the reV side but makes sense. No need to fix for now.

**Review by grantbuster — APPROVED** [review]

**Pull request merged**